# Esto es un codigo para calculos rapidos

In [ ]:
import numpy as np

def calcular_2theta(a, h, k, l, longitud_onda=1.5406):
    """
    Calcula el ángulo 2-theta para una estructura cúbica.
    
    Parámetros:
    a : float o numpy array
        Parámetro de red de la celda unitaria en Angstroms (Å).
    h, k, l : int
        Índices de Miller del plano cristalográfico.
    longitud_onda : float
        Longitud de onda de la fuente de rayos X en Angstroms.
        Por defecto es 1.5406 (Cu K-alpha).
        
    Retorna:
    float o numpy array con el valor del ángulo 2-theta en grados.
    """
    # Calcular el término de la raíz cuadrada de los índices de Miller
    raiz_hkl = np.sqrt(h**2 + k**2 + l**2)
    
    # Calcular el argumento del arcoseno: (lambda * raiz) / (2 * a)
    argumento_arcsin = (longitud_onda * raiz_hkl) / (2 * a)
    
    # Validar que el argumento sea válido para la función arcoseno (<= 1)
    # Si es mayor a 1, físicamente no hay difracción posible para esa configuración
    if np.any(argumento_arcsin > 1):
        raise ValueError("El parámetro de red 'a' es demasiado pequeño y no cumple la Ley de Bragg para este plano.")
    
    # Calcular theta en radianes y convertirlo a grados
    theta_rad = np.arcsin(argumento_arcsin)
    theta_deg = np.degrees(theta_rad)
    
    # El difractograma muestra 2-theta, así que multiplicamos por 2
    return 2 * theta_deg

# ==========================================
# Ejemplo de aplicación para el plano (002)
# ==========================================

# 1. Evaluando un solo parámetro de red (ej. a = 5.43 Angstroms para el Silicio)

a_InN = 4.98
a_GaN = 4.52

m914_In = [ 12.61, 8.62, 8.58 ]
m925_In = [ 10.9, 11.85, 12.43 ]
m930_In = [ 25.38, 18.3, 15.6 ]
m931_In = [ 26.3, 17.34, 15.14 ]

# Porcentaje de In en InGaN



for x_In in m931_In:
    x = x_In / 100

    a_InGaN = x * a_InN + (1 - x) * a_GaN


    dos_theta_002 = calcular_2theta(a=a_InGaN, h=0, k=0, l=2)

    print(f"Cálculo individual:")
    print(f"Para a = {a_InGaN} Å y el plano (002), el pico aparecerá en 2θ = {dos_theta_002:.2f}°\n")




Cálculo individual:
Para a = 4.64098 Å y el plano (002), el pico aparecerá en 2θ = 38.78°

Cálculo individual:
Para a = 4.5997639999999995 Å y el plano (002), el pico aparecerá en 2θ = 39.14°

Cálculo individual:
Para a = 4.589644 Å y el plano (002), el pico aparecerá en 2θ = 39.23°



### m914
Cálculo individual:
Para a = 4.578005999999999 Å y el plano (002), el pico aparecerá en 2θ = 39.33°

Cálculo individual:
Para a = 4.559651999999999 Å y el plano (002), el pico aparecerá en 2θ = 39.50°

Cálculo individual:
Para a = 4.559468 Å y el plano (002), el pico aparecerá en 2θ = 39.50°


### m925
Cálculo individual:
Para a = 4.570139999999999 Å y el plano (002), el pico aparecerá en 2θ = 39.40°

Cálculo individual:
Para a = 4.574509999999999 Å y el plano (002), el pico aparecerá en 2θ = 39.36°

Cálculo individual:
Para a = 4.577178 Å y el plano (002), el pico aparecerá en 2θ = 39.34°

### m930
Cálculo individual:
Para a = 4.636748 Å y el plano (002), el pico aparecerá en 2θ = 38.81°

Cálculo individual:
Para a = 4.6041799999999995 Å y el plano (002), el pico aparecerá en 2θ = 39.10°

Cálculo individual:
Para a = 4.59176 Å y el plano (002), el pico aparecerá en 2θ = 39.21°

### m931

Cálculo individual:
Para a = 4.64098 Å y el plano (002), el pico aparecerá en 2θ = 38.78°

Cálculo individual:
Para a = 4.5997639999999995 Å y el plano (002), el pico aparecerá en 2θ = 39.14°

Cálculo individual:
Para a = 4.589644 Å y el plano (002), el pico aparecerá en 2θ = 39.23°

In [10]:
m914_In = [ 12.61, 8.62, 8.58 ]
m925_In = [ 10.9, 11.85, 12.43 ]
m930_In = [ 25.38, 18.3, 15.6 ]
m931_In = [ 26.3, 17.34, 15.14 ]
